py/embedded-server/deephaven_server/cli.py

```diff
@@ -0,0 +1,79 @@
+import click
+import signal
+import sys
+import webbrowser
+
+from .server import Server
+
+@click.group()
+def cli():
+    """
+    Command-line interface entry point for the Deephaven embedded server application.
+    Accepts a command to start the application and the arguments to use.
+    """
+    pass
+
+
+@cli.command()
+@click.option("--host", default=None, help="The host to bind to.")
+@click.option("--port", default=None, type=int, help="The port to bind to.")
+@click.option(
+    "--key",
+    default=None,
+    help="Start the server in pre-shared key mode. If the key is empty, a random key will be used. Cannot be used if anonymous is specified.",
+)
+@click.option(
+    "--anonymous",
+    is_flag=True,
+    help="Start the server in anonymous mode. Cannot be used if key is specified.",
+)
+@click.option("--jvm-args", default=None, help="The JVM arguments to use.")
+def server(host, port, key, anonymous, jvm_args):
+    """
+    Start the Deephaven server.
+    """
+    click.echo("Starting Deephaven server...")
+
+    if anonymous and key is not None:
+        raise click.ClickException("Cannot specify both --anonymous and --key")
+
+    if jvm_args is None:
+        jvm_args = ""
+    jvm_args = jvm_args.split()
+
+    if anonymous:
+        jvm_args += [f"-DAuthHandlers=io.deephaven.auth.AnonymousAuthenticationHandler"]
+    elif key is not None:
+        jvm_args += [
+            f"-DAuthHandlers=io.deephaven.authentication.psk.PskAuthenticationHandler"
+        ]
+        # We are relying on server to set random key otherwise
+        if key:
+            jvm_args += [f"-Dauthentication.psk={key}"]
+
+    s = Server(host=host, port=port, jvm_args=jvm_args)
+    s.start()
+
+    target_url_or_default = s.server_config.target_url_or_default
+    authentication_type = None
+    authentication_url = target_url_or_default
+    for authentication_handler in s.authentication_handlers:
+        authentication_urls = authentication_handler.urls(target_url_or_default)
+        if authentication_urls:
+            authentication_type = authentication_handler.auth_type
+            authentication_url = authentication_urls[0]
+            break
```

Comment:
@devinrsmith / @niloc132 any reason we don't just have a method on `EmbeddedServer` that does all this song and dance to return all the auth URLs?
I guess we could have multiple authentication handlers, but I feel like 90% of the time it's just going to be one auth handler and we'll have one default URL. Doesn't matter much though.

Reply:
That's the route I originally took when creating this feature; but it became clear that I was trying to service this one specific use case as opposed to being more library-like and presenting the components themselves.

I could see an argument for moving this logic into the _python_ side `Server`; I don't necessarily think the java side `EmbeddedServer` is the correct place. Once you elevate it to `Server` though, it becomes a public API and something we need to treat as a library feature more generally.

Reply:
With that said, this cli _is_ a public API - once we put it out there, we can't break it (without being the kind of jerk that would break any other API).